### PR TITLE
Autoupdates: add more data to the jetpack_updates option

### PR DESCRIPTION
**Pull request by roccotripaldi**
The jetpack_updates options should also contain the contents of
the following core transients: update_core, update_plugins, update_themes.

jetpack_updates should be synced more regularly, not just when a transient is updated.

and jetpack_updates should be exposed on the /sites/$site, and /sites/me endpoints.

### CODE REVIEW

File: class.jetpack-autoupdate.php [hunk 1]
```diff
@@ -111,11 +111,14 @@ function save_update_data() {
...
 		// Stores the individual update counts as well as the total count.
 		if ( isset( $update_data['counts'] ) ) {
-			$updates = $update_data['counts'];
+			$updates['counts'] = $update_data['counts'];
 		}
 
 		// Stores the current version of WordPress.
 		$updates['wp_version'] = $wp_version;
+		$updates['update_core'] = get_site_transient( 'update_core' );
+		$updates['update_plugins'] = get_site_transient( 'update_core' );
```
Comment (lezama) [inline class.jetpack-autoupdate.php]: `update_core` -> `update_plugins`

**Comment by roccotripaldi** [comment]
This is working well for me, but I would like some more eyes on it.
The only concern I have is that we are adding a large chunk of data to
/sites/$site and /sites/mine

Here is what will be added:
```
Array
        (
            [counts] => Array
                (
                    [plugins] => 0
                    [themes] => 0
                    [wordpress] => 0
                    [translations] => 0
                    [total] => 0
                )

            [wp_version] => 4.0.7
            [update_core] => stdClass Object
                (
                    [updates] => Array
                        (
                            [0] => stdClass Object
                                (
                                    [response] => upgrade
                                    [download] => https://downloads.wordpress.org/release/wordpress-4.2.4.zip
                                    [locale] => en_US
                                    [packages] => stdClass Object
                                        (
                                            [full] => https://downloads.wordpress.org/release/wordpress-4.2.4.zip
                                            [no_content] => https://downloads.wordpress.org/release/wordpress-4.2.4-no-content.zip
                                            [new_bundled] => https://downloads.wordpress.org/release/wordpress-4.2.4-new-bundled.zip
                                            [partial] => 
                                            [rollback] => 
                                        )

                                    [current] => 4.2.4
                                    [version] => 4.2.4
                                    [php_version] => 5.2.4
                                    [mysql_version] => 5.0
                                    [new_bundled] => 4.1
                                    [partial_version] => 
                                )

                            [1] => stdClass Object
                                (
                                    [response] => autoupdate
                                    [download] => https://downloads.wordpress.org/release/wordpress-4.2.4.zip
                                    [locale] => en_US
                                    [packages] => stdClass Object
                                        (
                                            [full] => https://downloads.wordpress.org/release/wordpress-4.2.4.zip
                                            [no_content] => https://downloads.wordpress.org/release/wordpress-4.2.4-no-content.zip
                                            [new_bundled] => https://downloads.wordpress.org/release/wordpress-4.2.4-new-bundled.zip
                                            [partial] => 
                                            [rollback] => 
                                        )

                                    [current] => 4.2.4
                                    [version] => 4.2.4
                                    [php_version] => 5.2.4
                                    [mysql_version] => 5.0
                                    [new_bundled] => 4.1
                                    [partial_version] => 
                                    [support_email] => updatehelp42@wordpress.org
                                )

                            [2] => stdClass Object
                                (
                                    [response] => autoupdate
                                    [download] => https://downloads.wordpress.org/release/wordpress-4.1.7.zip
                                    [locale] => en_US
                                    [packages] => stdClass Object
                                        (
                                            [full] => https://downloads.wordpress.org/release/wordpress-4.1.7.zip
                                            [no_content] => https://downloads.wordpress.org/release/wordpress-4.1.7-no-content.zip
                                            [new_bundled] => https://downloads.wordpress.org/release/wordpress-4.1.7-new-bundled.zip
                                            [partial] => 
                                            [rollback] => 
                                        )

                                    [current] => 4.1.7
                                    [version] => 4.1.7
                                    [php_version] => 5.2.4
                                    [mysql_version] => 5.0
                                    [new_bundled] => 4.1
                                    [partial_version] => 
                                    [support_email] => updatehelp42@wordpress.org
                                )

                        )

                    [last_checked] => 1438849396
                    [version_checked] => 4.0.7
                    [translations] => Array
                        (
                        )

                )

            [update_plugins] => stdClass Object
                (
                    [last_checked] => 1438866689
                    [checked] => Array
                        (
                            [akismet/akismet.php] => 3.1.1
                            [debug-bar/debug-bar.php] => 0.8.1
                            [developer/developer.php] => 1.2.5
                            [hello.php] => 1.6
                            [jetpack/jetpack.php] => 3.7.0-dev
                            [log-deprecated-notices/log-deprecated-notices.php] => 0.3
                            [log-viewer/log-viewer.php] => 14.05.01
                            [tester.php] => 1.0
                            [jetpack-test-automatic-updates/jetpack-test-automattic-updates.php] => 1.0
                        )

                    [response] => Array
                        (
                            [akismet/akismet.php] => stdClass Object
                                (
                                    [id] => 15
                                    [slug] => akismet
                                    [plugin] => akismet/akismet.php
                                    [new_version] => 3.1.3
                                    [url] => https://wordpress.org/plugins/akismet/
                                    [package] => https://downloads.wordpress.org/plugin/akismet.3.1.3.zip
                                )

                            [debug-bar/debug-bar.php] => stdClass Object
                                (
                                    [id] => 18561
                                    [slug] => debug-bar
                                    [plugin] => debug-bar/debug-bar.php
                                    [new_version] => 0.8.2
                                    [url] => https://wordpress.org/plugins/debug-bar/
                                    [package] => https://downloads.wordpress.org/plugin/debug-bar.0.8.2.zip
                                    [upgrade_notice] => Updated to handle a new deprecated message in WordPress 4.0.
                                )

                            [log-viewer/log-viewer.php] => stdClass Object
                                (
                                    [id] => 35109
                                    [slug] => log-viewer
                                    [plugin] => log-viewer/log-viewer.php
                                    [new_version] => 14.05.04
                                    [url] => https://wordpress.org/plugins/log-viewer/
                                    [package] => https://downloads.wordpress.org/plugin/log-viewer.14.05.04.zip
                                )

                        )

                    [translations] => Array
                        (
                        )

                    [no_update] => Array
                        (
                            [developer/developer.php] => stdClass Object
                                (
                                    [id] => 24908
                                    [slug] => developer
                                    [plugin] => developer/developer.php
                                    [new_version] => 1.2.5
                                    [url] => https://wordpress.org/plugins/developer/
                                    [package] => https://downloads.wordpress.org/plugin/developer.1.2.5.zip
                                )

                            [hello.php] => stdClass Object
                                (
                                    [id] => 3564
                                    [slug] => hello-dolly
                                    [plugin] => hello.php
                                    [new_version] => 1.6
                                    [url] => https://wordpress.org/plugins/hello-dolly/
                                    [package] => https://downloads.wordpress.org/plugin/hello-dolly.1.6.zip
                                )

                            [jetpack/jetpack.php] => stdClass Object
                                (
                                    [id] => 20101
                                    [slug] => jetpack
                                    [plugin] => jetpack/jetpack.php
                                    [new_version] => 3.6.1
                                    [url] => https://wordpress.org/plugins/jetpack/
                                    [package] => https://downloads.wordpress.org/plugin/jetpack.3.6.1.zip
                                )

                            [log-deprecated-notices/log-deprecated-notices.php] => stdClass Object
                                (
                                    [id] => 15700
                                    [slug] => log-deprecated-notices
                                    [plugin] => log-deprecated-notices/log-deprecated-notices.php
                                    [new_version] => 0.3
                                    [url] => https://wordpress.org/plugins/log-deprecated-notices/
                                    [package] => https://downloads.wordpress.org/plugin/log-deprecated-notices.0.3.zip
                                    [upgrade_notice] => Updated to handle a new deprecated message in WordPress 4.0.
                                )

                        )

                )

            [update_themes] => stdClass Object
                (
                    [last_checked] => 1438849397
                    [checked] => Array
                        (
                            [confit-wpcom] => 1.2
                            [twentyeleven] => 2.1
                            [twentyfifteen] => 1.1
                            [twentyfourteen] => 1.4
                            [twentyten] => 1.9
                            [twentythirteen] => 1.5
                            [twentytwelve] => 1.7
                        )

                    [response] => Array
                        (
                            [twentyfifteen] => Array
                                (
                                    [theme] => twentyfifteen
                                    [new_version] => 1.2
                                    [url] => https://wordpress.org/themes/twentyfifteen/
                                    [package] => https://downloads.wordpress.org/theme/twentyfifteen.1.2.zip
                                )

                        )

                    [translations] => Array
                        (
                        )

                )

            [site_is_version_controlled] => 
        )
```

If that seems like it is too much, we could omit the update_core, update_plugins, and update_themes transient.  We might use that data in the future, but we are not currently using it.
@koke mentioned plans to send push notifications when we discover a plugin is out of date.

File: class.jetpack-autoupdate.php [hunk 2]
```diff
@@ -42,11 +42,15 @@ private function __construct() {
...
 		}
 
-		// Anytime WordPress saves update data, we'll want to update our Jetpack option as well.
 		if ( is_main_site() ) {
+			// Anytime WordPress saves update data, we'll want to update our Jetpack option as well.
 			add_action( 'set_site_transient_update_plugins', array( $this, 'save_update_data' ) );
 			add_action( 'set_site_transient_update_themes', array( $this, 'save_update_data' ) );
 			add_action( 'set_site_transient_update_core', array( $this, 'save_update_data' ) );
+			// Anytime a connection to jetpack is made, sync the update data
+			add_action( 'update_option_jetpack_blog_token', array( $this, 'save_update_data' ) );
```
Comment (lezama) [inline class.jetpack-autoupdate.php]: Good idea :) but this won't work as `blog_token` is stored inside the option `jetpack_private_options`. I would add a new hook `do_action( 'jetpack_site_registered' )` to the `register` method inside `class.jetpack.php`.

File: json-endpoints/class.wpcom-json-api-get-site-endpoint.php [hunk 3]
```diff
@@ -314,7 +314,8 @@ public function build_current_site_response( ) {
 					// If the sync happends right then we should be able to see that we are not dealing with a network site
 					$response['options']['is_multi_network'] = (bool) get_option( 'jetpack_is_main_network', true  );
 					$response['options']['is_multi_site'] = (bool) get_option( 'jetpack_is_multi_site', true );
-
+					// Show available updates
+					$response['options']['updates'] = get_option( 'jetpack_updates' );
```
Comment (lezama) [inline json-endpoints/class.wpcom-json-api-get-site-endpoint.php]: We should use the `Jetpack_Options` methods in order to handle Jetpack options.

**Comment by roccotripaldi** [comment]
Thanks for the feedback @lezama - changes are in.

Question: Should we output the jetpack_updates option on sites/$site/updates ?

**Comment by koke** [comment]
> The only concern I have is that we are adding a large chunk of data to /sites/$site and /sites/mine

To me it feels like a lot to expose. I think we want that data synced to wp.com so it's readily available, but maybe only expose update counts in the API, unless it's a more specific `updates` endpoint.

**Comment by roccotripaldi** [comment]
> maybe only expose update counts in the API,

Perfect. Thanks @koke 

I think this one is ready to go in.

File: json-endpoints/class.wpcom-json-api-get-site-endpoint.php [hunk 4]
```diff
@@ -315,6 +315,8 @@ public function build_current_site_response( ) {
 					$response['options']['is_multi_network'] = (bool) get_option( 'jetpack_is_main_network', true  );
 					$response['options']['is_multi_site'] = (bool) get_option( 'jetpack_is_multi_site', true );
 
+					// Show available updates.
+					$response['options']['updates'] = Jetpack_Options::get_option( 'updates' );
```
Comment (enejb) [inline json-endpoints/class.wpcom-json-api-get-site-endpoint.php]: I think this part will change. I would better to omit it from and then use what ever is coming from .com instead.

**Comment by enejb** [comment]
Looking though the code it looks like if a person that doesn't have 'current_user_can( 'update_plugins' )' privileges registers Jetpack. Data will not get synced on registration of Jetpack. 
See https://developer.wordpress.org/reference/functions/wp_get_update_data/

**Comment by roccotripaldi** [comment]
@enejb : do you think this makes more sense?
https://github.com/Automattic/jetpack/pull/2429/files#diff-ac54787f5c8bcac1491098d8fb5a8d18R124

If the current user cannot manage plugins, there is not much we can do to sync any useful data. We should at least bail to avoid syncing incomplete data.

**Comment by enejb** [comment]
@roccotripaldi I think that makes sense. 
But we should also check for `update_themes` and `update_core` here as well. 
Since not having them would result in incomplete data as well.

**Comment by zinigor** [comment]
Rebased, read the changes, tested, works fine!

**Pull request merged**